src/main/kotlin/org/rust/lang/core/resolve/indexes/RsTypeAliasIndex.kt

```diff
@@ -0,0 +1,81 @@
+import org.rust.lang.core.psi.ext.containingCargoPackage
+import org.rust.lang.core.psi.ext.owner
+import org.rust.lang.core.psi.isValidProjectMember
+import org.rust.lang.core.resolve.RsProcessor
+import org.rust.lang.core.stubs.RsFileStub
+import org.rust.lang.core.stubs.RsTypeAliasStub
+import org.rust.lang.core.types.TyFingerprint
+import org.rust.openapiext.getElements
+
+class RsTypeAliasIndex : AbstractStubIndex<TyFingerprint, RsTypeAlias>() {
+    override fun getVersion(): Int = RsFileStub.Type.stubVersion
+    override fun getKey(): StubIndexKey<TyFingerprint, RsTypeAlias> = KEY
+    override fun getKeyDescriptor(): KeyDescriptor<TyFingerprint> = TyFingerprint.KeyDescriptor
+
+    companion object {
+        fun findPotentialAliases(
+            project: Project,
+            tyf: TyFingerprint,
+            processor: RsProcessor<RsTypeAlias>
+        ): Boolean {
+            // Note that `getElements` is intentionally used with intermediate collection instead of
+            // `StubIndex.processElements` in order to simplify profiling
+            val aliases = getElements(KEY, tyf, project, RsWithMacrosProjectScope(project))
+                .filter { it.owner is RsAbstractableOwner.Free && it.isValidProjectMember }
+
+            // This is basically a hack to make some crates (winapi 0.2) work in a reasonable amount of time.
+            // If the number of aliases exceeds the threshold, we prefer ones from stdlib and a workspace over
+            // aliases from dependencies
+            val filteredAliases = if (aliases.size <= ALIAS_COUNT_THRESHOLD) {
+                aliases
+            } else {
+                val stdlibAliases = mutableListOf<RsTypeAlias>()
+                val workspaceAliases = mutableListOf<RsTypeAlias>()
+                for (alias in aliases) {
+                    val packageOrigin = alias.containingCargoPackage?.origin ?: continue
+                    when (packageOrigin) {
+                        PackageOrigin.STDLIB -> stdlibAliases += alias
+                        PackageOrigin.WORKSPACE -> workspaceAliases += alias
+                        else -> Unit
+                    }
+                }
+                when {
+                    stdlibAliases.size + workspaceAliases.size <= ALIAS_COUNT_THRESHOLD -> {
+                        stdlibAliases + workspaceAliases
+                    }
+                    stdlibAliases.size <= ALIAS_COUNT_THRESHOLD -> stdlibAliases
+                    else -> return false
+                }
+            }
+
+            return filteredAliases.any { processor(it) }
+        }
+
+        fun index(stub: RsTypeAliasStub, sink: IndexSink) {
+            val alias = stub.psi
+            val typeRef = alias.typeReference ?: return
+            TyFingerprint.create(typeRef, emptyList())
+                .forEach { sink.occurrence(KEY, it) }
+        }
+
+        private const val ALIAS_COUNT_THRESHOLD = 10
```

Comment:
Maybe it's better to add registry key with the same default value to have the opportunity to fix some local user issues because of too small/too large threshold?

Reply:
Sounds reasonable. Done